### PR TITLE
buildGraalvm: fix build on x86_64-darwin

### PR DESCRIPTION
**Pull request by nathanregner**
Related: https://github.com/NixOS/nixpkgs/issues/354059

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/compilers/graalvm/community-edition/buildGraalvm.nix [hunk 1]
```diff
@@ -107,7 +110,10 @@ let
       ++ lib.optional stdenv.hostPlatform.isLinux autoPatchelfHook;
 
     propagatedBuildInputs = [ setJavaClassPath zlib ]
-      ++ lib.optional stdenv.hostPlatform.isDarwin darwin.apple_sdk_11_0.frameworks.Foundation;
+      ++ lib.optionals (stdenv.hostPlatform.isDarwin && lib.versionOlder apple-sdk.version "11") [
```
Comment (nathanregner) [inline pkgs/development/compilers/graalvm/community-edition/buildGraalvm.nix:113]: Stolen from here: https://github.com/NixOS/nixpkgs/pull/351004#discussion_r1815645876

**Comment by nathanregner** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 360254`

---
### `x86_64-linux`

---
### `aarch64-linux`

---
### `x86_64-darwin`
<details>
  <summary>:fast_forward: 3 packages marked as broken and skipped:</summary>
  <ul>
    <li>dapl-native</li>
    <li>dbqn-native</li>
    <li>obb</li>
  </ul>
</details>
<details>
  <summary>:x: 3 packages failed to build:</summary>
  <ul>
    <li>certificate-ripper</li>
    <li>graalvmCEPackages.graalnodejs</li>
    <li>scala-update</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 17 packages built:</summary>
  <ul>
    <li>babashka</li>
    <li>babashka-unwrapped</li>
    <li>bbin</li>
    <li>clj-kondo</li>
    <li>cljfmt</li>
    <li>clojure-lsp</li>
    <li>cq</li>
    <li>graalvm-ce</li>
    <li>graalvmCEPackages.graaljs</li>
    <li>graalvmCEPackages.graalpy</li>
    <li>graalvmCEPackages.truffleruby</li>
    <li>jet</li>
    <li>kdePackages.stdenv</li>
    <li>neil</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>yamlscript</li>
    <li>zprint</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 3 packages marked as broken and skipped:</summary>
  <ul>
    <li>dapl-native</li>
    <li>dbqn-native</li>
    <li>obb</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 16 packages built:</summary>
  <ul>
    <li>babashka</li>
    <li>babashka-unwrapped</li>
    <li>bbin</li>
    <li>certificate-ripper</li>
    <li>clj-kondo</li>
    <li>cljfmt</li>
    <li>clojure-lsp</li>
    <li>cq</li>
    <li>graalvm-ce</li>
    <li>jet</li>
    <li>kdePackages.stdenv</li>
    <li>neil</li>
    <li>scala-update</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>yamlscript</li>
    <li>zprint</li>
  </ul>
</details>

File: pkgs/development/compilers/graalvm/community-edition/buildGraalvm.nix (hunk 1 above)
Comment (paparodeo) [reply pkgs/development/compilers/graalvm/community-edition/buildGraalvm.nix:113]: so it turns out that the ` lib.versionOlder apple-sdk.version "11"`  logic is unnecessary. The build will use the highest sdk. so if some other package propagates `apple-sdk_12` then `apple-sdk_12` will be used even if we add `apple-sdk_11` here. or if `apple-sdk` ends up being version 12, we'll use version 12 even if we specify `apple-sdk_11` here.

```suggestion
      ++ lib.optionals stdenv.hostPlatform.isDarwin [
```

also note that both the Darwins (aarch64/x86_64) will be using the same sdk version (11) soonish (the change is in staging but maybe a few weeks before it makes its way to master)

File: pkgs/development/compilers/graalvm/community-edition/buildGraalvm.nix [hunk 2]
```diff
@@ -118,13 +120,23 @@ let
...
         cLibsAsFlags = (map (l: "--add-flags '-H:CLibraryPath=${l}/lib'") cLibs);
         preservedNixVariables = [
+          "-EDEVELOPER_DIR"
+          "-EDEVELOPER_DIR_FOR_BUILD"
+          "-EDEVELOPER_DIR_FOR_TARGET"
           "-ELOCALE_ARCHIVE"
+          "-EMACOSX_DEPLOYMENT_TARGET"
+          "-EMACOSX_DEPLOYMENT_TARGET_FOR_BUILD"
+          "-EMACOSX_DEPLOYMENT_TARGET_FOR_TARGET"
+          "-ENIX_APPLE_SDK_VERSION"
```
Comment (thiagokokada) [inline pkgs/development/compilers/graalvm/community-edition/buildGraalvm.nix:139]: Please add those variables behind a `lib.optionals stdenv.hostPlarform.isDarwin`.

File: pkgs/development/compilers/graalvm/community-edition/buildGraalvm.nix [hunk 3]
```diff
@@ -118,13 +120,23 @@ let
...
+      ]);
 
     postInstall =
       let
         cLibsAsFlags = (map (l: "--add-flags '-H:CLibraryPath=${l}/lib'") cLibs);
         preservedNixVariables = [
+          "-EDEVELOPER_DIR"
+          "-EDEVELOPER_DIR_FOR_BUILD"
+          "-EDEVELOPER_DIR_FOR_TARGET"
           "-ELOCALE_ARCHIVE"
```
Comment (thiagokokada) [inline pkgs/development/compilers/graalvm/community-edition/buildGraalvm.nix:135]: I am almost sure that this is Linux only, so could you move this to `lib.optionals stdenv.hostPlatform.isLinux`?

**Comment by thiagokokada** [comment]
@ofborg build babashka clojure-lsp

**Comment by kenota** [comment]
Hey @nathanregner , do you mind backporting this to nixos-24.11 release stable? I was trying to update my darwing setup today and faced this issue with building graalvm. 

According to the [https://github.com/NixOS/nixpkgs/blob/d88425e91ca77d6041fd0c1801012bcbe0fae56b/CONTRIBUTING.md#backporting-changes](Backporting section) you just need to add label `backport release-24.11` and the github actions will create the backport even though this PR is merged.

**Comment by FliegendeWurst** [comment]
Hydra confirms the same failure on stable https://hydra.nixos.org/build/280282119

**Pull request merged**